[Feat] MyPage UI 개발

kimyounghee425 opened the pull request:
## Summary

<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve?
-->
- 마이페이지 화면 구현
- List 공통컴포넌트 개발
- useModal 훅 개발

## Detail

- List 컴포넌트 만들었습니다. 사용법 주석 해당 디렉토리 내의 index 에 있음

- shared/hooks/modal 안에 useModal 훅 만들었습니다.
openHorizontalModal : 가로 버튼 모달
openVerticalModal : 세로 버튼 모달
openImageModal : 이미지 모달
-> LiveRecordPage 모달 구현 부분 리팩터링 완료

### 추가 작업
회원가입 시 context 에 유저정보 저장
앱 렌더링 전에 msw 먼저 실행하게 수정
마이페이지 msw 작업 완료

현재 기록중 페이지로 push 할 때 recordId: "1" 로 하드코딩
수정 예정

https://github.com/user-attachments/assets/c236115e-155f-4f0e-9c3e-e244ad669ddc


TODO : 응원 팀 변경, 닉네임 변경 기능 만들기



## How did you test this change?

<!--
  Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface.
  How exactly did you verify that your PR solves the issue you wanted to solve?
  If you leave this empty, your PR will very likely be closed.
-->

kimyounghee425 committed:
✨ [feat] mypage 틀 생성

kimyounghee425 committed:
✨ [feat] List 컴포넌트 제작

kimyounghee425 committed:
📝 [docs] 주석 작성

kimyounghee425 committed:
✨ [feat] MyPage 틀 만들기, MyPageList 개발

kimyounghee425 committed:
✨ [feat] useModal 훅 생성 (가로버튼 모달만 추가)

kimyounghee425 committed:
✨ [feat] 모달 3종 훅으로 분리, LiveRecordPage 리팩터링

kimyounghee425 committed:
✨ [feat] MyPage 에 모달 작업

kimyounghee425 committed:
📝 [docs] useModal 주석 추가

kimyounghee425 committed:
Merge branch 'develop' into feat/mypage

kimyounghee425 committed:
🔀 Merge develop

kych0912 commented:
api연결은 혹시 언제하시나요? 경기중 페이지에서도요!

kimyounghee425 commented:
> api연결은 혹시 언제하시나요? 경기중 페이지에서도요!

일단 남은 뷰 (응원 팀 변경 / 닉네임 변경) 만든 후 한번에 할 것 같습니다
오늘 리뷰하면서 msw 사용법을 조금 이해한거같아서요;;
남은 뷰를 오늘 하고 api 연동을 일요일날 하지 않을까 싶어요!!!!!

kych0912 commented on apps/web/src/pages/live-recording/ui/LiveRecordPage.tsx:
```diff
@@ -36,91 +37,45 @@ const LiveRecordPageInner: ActivityComponentType = () => {
   const leavePage = () => {
     setTimeout(() => {
-      replace('Login', {})
+      replace('My', {})
```
MyPage로 이동하나요?

kimyounghee425 replied:
옙. 근데 이거 테스트용이였습니다

kych0912 commented on apps/web/src/pages/live-recording/ui/LiveRecordPage.tsx:
```diff
@@ -5,13 +5,14 @@ import { RecordingCard } from '@/features/record/ui/RecordingCard'
 import { useFlow } from '@stackflow/react/future'
-import { OverlayModal } from '@/shared/ui/common/OverlayModal'
-import { OverlayProvider, useOverlay } from '@/hooks/useOverlay'
+import { OverlayProvider } from '@/hooks/useOverlay'
```
OverlayProvider는 app에서 렌더링 해도 괜찮습니다! 필요한 곳에서 렌더링하는것보단 귀찮지않게 app layer가 나을 것 같습니다!

kych0912 commented on apps/web/src/pages/mypage/ui/MyPage.tsx:
이부분은 feature별로 분리해주세요~ 비즈니스 로직이랑 ui가 같이 있는게 읽기가 어렵습니다!

예를들어 로그아웃, 탈퇴 로직은 features/mypage에 ui slice에서 auth라는 파일 안에 비즈니스 로직이랑 같이 적어주세요!

알림설정, 내정보도 이런 방식으로요!. 지금은 너무 혼잡되어있고, 이후 추가될 비즈니스 로직까지 생각하면 해당 함수가 너무 복잡해질거같네요!

kimyounghee425 replied:
저도 같은생각입니다 api 연동하며 수정예정입니다

kych0912 commented on apps/web/src/pages/mypage/ui/MyPageList.tsx:
```diff
@@ -0,0 +1,39 @@
+  onClick,
+  ...rest
+}: MyPageListProps) => {
+  const { push } = useFlow()
```
이건 왜...??

kych0912 commented on apps/web/src/pages/mypage/ui/MyPageList.tsx:
```diff
@@ -0,0 +1,39 @@
+      <>
+        <List.Root>
+          <List.Text>{children}</List.Text>
+          <List.Arrow onClick={onClick} />
```
onClick은 List component를 눌렀을 때 실행되어야하지 않나요?

kimyounghee425 replied:
일부러 Arrow 에만 달았는데 별롤까요??
아래 Arrow 를 버튼으로 래핑한 이유도 이거였습니다

kimyounghee425 replied:
생각해보니 이게 맞는것같습니다

kych0912 commented on apps/web/src/pages/mypage/ui/MyPageList.tsx:
```diff
@@ -0,0 +1,39 @@
+import type { ComponentProps, ReactNode } from 'react'
+import { cn } from '@/shared/lib/classnames'
+import { useFlow } from '@stackflow/react/future'
+import { List } from '@/shared/ui/common/List'
```
List 형태는 양 끝 좌,우 에만 컴포넌트가 들어가는데 합성 컴포넌트 방식보다는 props로 좌우 컴포넌트를 받는 형식이 나을 것 같습니다!

kych0912 replied:
왜냐하면 
```typescript
      <List.Root>
        <List.Text>{children}</List.Text>
        <List.SwitchItem />
      </List.Root>
```
이런 형태는 mypage에서만 사용하진 않을 수 있으니깐요! props component로 shared layer에 두어 좌,우 컴포넌트를 props로 커스텀하는게 나을 것 같습니다!

kych0912 commented on apps/web/src/shared/hooks/modal/useModal.tsx:
모달 사용하기 너무 편할것같습니다! 감사합니다~

kych0912 commented on apps/web/src/shared/ui/common/List/Arrow.tsx:
요건 버튼으로 만드신 이유가 있나요? 그냥 svg component사용하면 되지 않나요?

kimyounghee425 replied:
버튼으로 사용해서 버튼으로 감쌌습니다. 별론가요?!
Switch 도 마찬가지

kych0912 commented on apps/web/src/shared/ui/common/List/SwitchItem.tsx:
Switch도 왜...??

kych0912 commented on apps/web/src/shared/ui/common/List/index.ts:
합성 컴포넌트를 고집하실 필요는 없을 것 같습니다~

kych0912 reviewed (commented):
늦게 확인했네요 수고하셨습니다~ 전체적으로 잘 작성하신 것 같습니다! 대신 컴포넌트 구현 방식은 컴파운드 대신 다른 디자인 패턴을 고려해보시면 좋을 것 같습니다!

kimyounghee425 committed:
✨ [feat] 감정 기록 msw 작업

kimyounghee425 committed:
♻️ [fix] record id 값 받게 수정, post -> get 수정

kimyounghee425 committed:
✨ [feat] useMutation 작업

kimyounghee425 committed:
✨ [feat] msw 연동

kimyounghee425 committed:
✨ [feat] emotion 상태 서버 응답 기반으로 갱신

kimyounghee425 committed:
🔀 Merge develop

kimyounghee425 committed:
💄 [style] 기록중 페이지 배경 그라데이션 작업

kimyounghee425 committed:
♻️ [refactor] List 컴포넌트 단일컴포넌트로 수정

kimyounghee425 committed:
✨ [feat] 팀 변경 페이지 생성

kimyounghee425 committed:
✨ [feat] 이름 변경 페이지 생성

kimyounghee425 committed:
✨ [feat] 회원가입 onSuccess 시 캐싱 작업 중간 저장

kimyounghee425 committed:
🔀 Merge dev

kimyounghee425 committed:
✏️ [fix] lint 적용

kimyounghee425 committed:
✨ [feat] 회원가입시 get 핸들러 msw 연동 작업

kimyounghee425 committed:
✨ [feat] 팀 변경 페이지 초기화면 완성

기존의 팀 선택되게 props 로 전달

kimyounghee425 committed:
✨ [feat] 팀 변경 폼 onSubmit 핸들러  작업 제외 완료

kimyounghee425 committed:
✨ [feat] useUpdateMyInfoMutation 훅 작성 / 페이지 연동

개인정보 수정 페이지에 공통으로 사용할 훅 작성

kimyounghee425 committed:
🐛 [fix] 앱 렌더링 이전에 MSW 서비스워커 실행

kimyounghee425 committed:
✨ [feat] patch msw 추가

kimyounghee425 committed:
💚 [fix] lint 에러 수정, 빌드 완료

kimyounghee425 committed:
💚 Resolve build

kimyounghee425 committed:
💚 [chore] QueryWrapper 에 MeProvider 추가

kych0912 commented on apps/web/src/app/main.tsx:
```diff
@@ -3,13 +3,22 @@ import { createRoot } from 'react-dom/client'
-)
+  createRoot(document.getElementById('root')!).render(
+    <StrictMode>
+      <MeProvider>
```
App.tsx에서 Provider를 감싸주세요~ 여기는 진입접으로 렌더링 역할을 하고 있는 함수입니다!

kimyounghee425 replied:
아 이거 빌드에러 해결하려고 이거저거 해보다 한거..니깐 수정합니다.

kych0912 commented on apps/web/src/entities/auth/api/auth-get.ts:
```diff
@@ -1,9 +1,14 @@
     const response = await api.get(`/auth/nickname/${nickname}`)
     return response.json()
   },
+  me: async (): Promise<MeResponseDTO> => {
```
여긴 get method함수를 모아두는 곳이라 앞에 get동사를 앞에 붙여주세요~ 그리고 me라는 이름도 적합하지 않은 것 같습니다! session이나 다른 이름으로 바꿔주세요!

kimyounghee425 replied:
me -> user 변경
context 네이밍 session 으로 변경

kych0912 commented on apps/web/src/entities/auth/api/auth-patch.ts:
```diff
@@ -0,0 +1,13 @@
+} from '@/entities/auth/model/auth.type'
+
+export const authPatch = {
+  updateUserInfo: async ({ nickname, baseballTeam }: SignupRequestDTO) => {
```
patch method면 patchUserInfo 이런식으로 바꿔주세요!

kych0912 commented on apps/web/src/features/auth/hooks/useMyInfo.ts:
```diff
@@ -0,0 +1,10 @@
+
+import { authGet } from '@/entities/auth/api'
+
+export const useMyInfo = () => {
```
emotion.queries.ts 에서 createQueryKey로 생성한걸로 그냥 useQuery(emotions.record(recordId)) 로 사용하시면 됩니다~ 따로 훅 만드실 필요 없어요!

kych0912 replied:
쿼리 키 관리, 캐싱, api 관리 책임은 @lukemorales/query-key-factory 라이브러리에서 합니다!

kimyounghee425 replied:
아 이거 안쓰는건데 삭제를 깜빡했네요;;;;

kych0912 commented on apps/web/src/mocks/data/emotion.ts:
```diff
@@ -0,0 +1,49 @@
+    recentEmotion: 'POSITIVE',
+    defaultImageUrl:
+      'https://ballog-bucket.s3.ap-northeast-2.amazonaws.com/images/24b81566-62aa-4351-854c-a4a79c22050c.svg',
+  } as EmotionType,
```
타입 캐스팅은 되도록이면 안쓰시는게....

kych0912 commented on apps/web/src/features/record/hooks/usePostEmotion.ts:
👍

kych0912 commented on apps/web/src/features/record/hooks/usePostEmotion.ts:
```diff
@@ -0,0 +1,30 @@
+  const queryClient = useQueryClient()
+  const { setEmotionPercent } = useEmotionVote()
+
+  return useMutation<EmotionResponseDTO, Error, PostEmotionProps>({
```
이부분은 낙관적 업데이트를 고려해도 좋을 것 같습니다!

kych0912 commented on apps/web/src/mocks/data/me.ts:
```diff
@@ -0,0 +1,13 @@
+    baseballTeam: 'LG_TWINS',
+    isNewUser: false,
+    role: 'USER',
+  } as MeType,
```
여기도 타입 캐스팅...

kych0912 commented on apps/web/src/mocks/handlers/emotionHandlers.ts:
```diff
@@ -0,0 +1,45 @@
+  },
+)
+
+export const emotionHandlers = [getEmotionHandler, postEmotionHandler]
```
코드 컨벤션 때문에... export const recordHandlers = [
  http.get<never, RecordResponseDTO, ApiErrorMessage | RecordResponseDTO>(
    `${RECORD_API_PREFIX}`,
    () => {
      // 네트워크 지연 효과 추가
      delay(3000)
      
 이런식으로 바로 배열에 넣어주실 수 있나요?

kych0912 commented on apps/web/src/mocks/handlers/meHandlers.ts:
```diff
@@ -0,0 +1,44 @@
+import { me } from '@/mocks/data/me'
+import type { SignupRequestDTO } from '@/entities/auth/model/auth.type'
+
+const ME_API_PREFIX = `${import.meta.env.VITE_PUBLIC_API_URL}/api/v1/mypage`
```
여기까지 와서 보니깐.... ME보단 Session이라는 이름이 잘 어울릴 것 같습니다. Me는 정확히 어떤 의미인지 모르겠는데 Session이라는 이름으로 해당 유저의 세션을 나타내는게 좋을 것 같습니다!

kych0912 commented on apps/web/src/pages/auth/ui/NickNamePage.tsx:
```diff
@@ -11,21 +11,28 @@ import { type ExtendedKyHttpError } from '@/types/api/common'
-    onSuccess: (data) => {
+
+    onSuccess: async (data) => {
+      const user = await authGet.me()
```
여기서 Promise를 처리하는 것 보단, 전역 상태에서 Promise객체를 받고, MeContext를 사용하는 곳에서 use(MeContext)로 유저 정보를 받아 선언적으로 처리하는게 좋을 것 같습니다! 지금 상태면 기존 유저가 로그인 할 때에는 NickNamePage를 건너지 않아 context에 유저 정보가 안담깁니다!!!!!!!

kych0912 replied:
```typescript
type UserContextType = {
  userPromise: Promise<User | null>;
};

const UserContext = createContext<UserContextType | null>(null);

let userPromise = authGet.me();
  
<UserProvider userPromise={userPromise}>
...
</UserProvider>
```
```typescript
const { userPromise } = useMeContext();
const user = use(userPromise);
```
이런식으로 use를 사용하는 패턴이 있답니다.

kych0912 replied:
아니면 react-query랑 같이 사용하는것도 좋고 suspense 컴포넌트를 사용해도 좋습니다.
```typescript
const UserProvider = ({ children }) => {
  const userQuery = useQuery({
    queryKey: ['user'],
    queryFn: authGet.me,
    staleTime: 10 * 60 * 1000,
    retry: false, // 401시 재시도 안함
  })
  
  return (
    <UserContext.Provider value={userQuery}>
      {children}
    </UserContext.Provider>
  )
}

// 컴포넌트에서
const { data: user, isLoading } = useUser()
if (isLoading) return <HeaderSkeleton />
return user ? <UserMenu user={user} /> : <LoginButton />
```
아니면 App에서 Suspense
```typescript
function App() {
  return (
    <ErrorBoundary fallback={<ErrorPage />}>
      <Suspense fallback={<GlobalLoading />}>
        <UserProvider>
          <Router />
        </UserProvider>
      </Suspense>
    </ErrorBoundary>
  )
}
```

kych0912 replied:
그리고 아마 미들웨어로 protected route를 구현할거라, use패턴이 편할 것 같습니다~

kimyounghee425 replied:
meContext 안에 
```
  const refetchUser = async () => {
    try {
      const data = await authGet.getUser()
      setUser(data.data)
    } catch (error) {
      setUser(null)
      throw error
    }
  }

  useEffect(() => {
    refetchUser()
  }, [])
  ```
  
  이 부분으로 첫 진입시 무조건 정보 가져옵니다!

kych0912 replied:
이거 error발생하면 어디서 catch하나요?

kych0912 commented on apps/web/src/pages/mypage/ui/ChangeNickNamePage.tsx:
```diff
@@ -0,0 +1,55 @@
+        baseballTeam: me.baseballTeam,
+      },
+      {
+        onSuccess: () => {
```
여기서 나중에 toast도 띄워야하는데, 일단 머지후에 수정하시죠~ sonner라는 라이브러리를 설치했습니다!

kych0912 commented on apps/web/src/pages/mypage/ui/MyPage.tsx:
```diff
@@ -0,0 +1,103 @@
+          <List type="arrow">서비스 이용약관</List>
+        </div>
+
+        <div className="flex items-center gap-4 px-4 mt-4 mb-4">
```
로그아웃, 탈퇴는 따로 분리 안하시나요?

kimyounghee425 replied:
아 이거 아직 없습니다

kych0912 commented on apps/web/src/shared/contexts/meContext.tsx:
Session이라는 이름을 고려해보시는게....ㅎㅎ

kych0912 commented on apps/web/src/test/QueryWrapper.tsx:
👍

kych0912 reviewed (commented):
확인했습니다~ Me라는 이름대신 Session을 한번 고려해주세요! 그리고 use 패턴으로 유저 상태 관리하는게 좋을 것 같습니다!

저는 이런식으로 사용합니다!
- https://github.com/kych0912/saas-starter-ko/blob/main/apps/client/lib/auth/index.tsx
- https://github.com/kych0912/saas-starter-ko/blob/main/apps/client/app/%5Blng%5D/layout.tsx
지금은 CSR환경이라 SSR과 다르긴한데, 큰 맥락은 제가 단 comment처럼 이해해주세요!

kimyounghee425 committed:
♻️ [fix] me -> user, context 네이밍 변경

kimyounghee425 committed:
♻️ [fix] 리뷰 반영

kimyounghee425 committed:
:wq
Merge branch 'develop' of https://github.com/Central-MakeUs/ballog-front into develop

kimyounghee425 committed:
🔀 Merge develop
